terraform-plans/templates/github/tox.ini.tftpl

```diff
@@ -0,0 +1,103 @@
+# This file is centrally managed as a template file in https://github.com/canonical/solutions-engineering-automation
+# To update the file:
+# - Edit it in the canonical/solutions-engineering-automation repository.
+# - Open a PR with the changes.
+# - When the PR merges, the soleng-terraform bot will open a PR to the target repositories with the changes.
+
+[tox]
+skipsdist=True
+skip_missing_interpreters = True
+envlist = lint, unit
+
+[testenv]
+basepython = python3
+setenv =
+  PYTHONPATH = {toxinidir}:{toxinidir}/src/:{toxinidir}/reactive/:{toxinidir}/hooks/:{toxinidir}/lib/:{toxinidir}/actions:{toxinidir}/files/:{toxinidir}/files/plugins/
+  # avoid state written to file during tests - see https://github.com/juju/charm-helpers/blob/85dcbeaf63b0d0f38e8cb17825985460dc2cd02d/charmhelpers/core/unitdata.py#L179-L184
+  UNIT_STATE_DB = :memory:
+passenv = *
+
+[testenv:lint]
+commands =
+    black --check --diff --color .
+    isort --check --diff --color .
+    flake8
+    pylint --recursive=y .
+    mypy --install-types --non-interactive .
+deps =
+    black
+    colorama
+    flake8
+    flake8-colors
+    flake8-docstrings
+    flake8-import-order
+    flake8-pyproject
+    isort
+    mypy
+    pep8-naming
+    pylint
+    # so pylint and mypy can reason about the code
+    {[testenv:unit]deps}
+    {[testenv:func]deps}
+
+[testenv:reformat]
+commands =
+    black .
+    isort .
+deps =
+    black
+    isort
+
+[testenv:unit]
+setenv =
+    {[testenv]setenv}
+    COVERAGE_FILE = .coverage-unit
+%{ if unittest_type == "none" ~}
+allowlist_externals =
+    echo
+commands = echo "No unit tests, skipping."
+%{ else ~}
+%{ if unittest_type == "pytest" ~}
+commands = pytest {toxinidir}/tests/unit \
+   -v \
+   --cov \
+   --cov-report=term-missing \
+   --cov-report=html \
+   --cov-report=xml \
+   {posargs}
+
+deps =
+  pytest
+  pytest-cov
+  -r {toxinidir}/requirements.txt
+  -r {toxinidir}/tests/unit/requirements.txt
+%{ else ~}
+ERROR: invalid `unittest_type` value
+%{ endif ~}
+%{ endif ~}
+
+[testenv:func]
+setenv =
+    {[testenv]setenv}
+    COVERAGE_FILE = .coverage-func
+%{ if functest_type == "none" ~}
+allowlist_externals =
+    echo
+commands = echo "No func tests, skipping."
+%{ else ~}
+deps =
+  pytest
+  pytest-cov
+  pytest-operator
+  -r {toxinidir}/requirements.txt
+  -r {toxinidir}/tests/functional/requirements.txt
+%{ if functest_type == "zaza" ~}
+changedir = {toxinidir}/tests/functional
+commands = functest-run-suite {posargs:--keep-model -v}
+%{ else ~}%{ if functest_type == "pytest" ~}
+commands = pytest {toxinidir}/tests/functional {posargs:-v}
```

Comment:
To say the truth, I don't know if we have a project that generates coverage for functional tests. However, I think we can start be ready for that. E.g: COU we can possibly use the functional tests with the python package instead of the snap to have some coverage.

With that said, I think we can use the same arguments of unit test to generate coverage.

Reply:
Ah agreed, that would be good. And we can also get coverage from the test .py files themselves, which is also important. :+1: